### PR TITLE
[GraphQL/Data] Loosen restriction on Boxed Queries

### PR DESCRIPTION
**Pull request by amnn**
## Description

Now the `DbConnection` (previously `QueryExecutor`) no longer needs to be given a boxed query.  This unblocks use of union queries, which we will likely need to support consistent object reads.

## Test Plan

Behaviour preserving refactor, passing all unit and E2E tests:

```
sui-graphql-rpc$ cargo nextest run
sui-graphql-e2e-tests$ cargo nextest run -j 1 --features pg_integration
```

## Stack 

- #15661 
- #15695

### CODE REVIEW

**Review by stefan-mysten — APPROVED** [review]

File: crates/sui-graphql-rpc/src/data.rs [hunk 1]
```diff
@@ -71,33 +67,29 @@ pub(crate) trait DbConnection {
...
-    fn result<Q, ST, QS, GB, U>(&mut self, query: Q) -> QueryResult<U>
+    fn result<Q, U>(&mut self, query: impl Fn() -> Q) -> QueryResult<U>
     where
-        Q: Fn() -> Query_<ST, QS, Self::Backend, GB>,
-        Query_<ST, QS, Self::Backend, GB>: LoadQuery<'static, Self::Connection, U>,
-        Query_<ST, QS, Self::Backend, GB>: QueryFragment<Self::Backend>,
-        QS: QuerySource;
+        Q: diesel::query_builder::Query,
+        Q: LoadQuery<'static, Self::Connection, U>,
+        Q: QueryId + QueryFragment<Self::Backend>;
```
Comment (wlmyng) [inline crates/sui-graphql-rpc/src/data.rs:76]: I see, looks like I was using the wrong syntax when trying this on my own - I was trying to Fn() -> dyn Trait, but this is inaccurate since the closure will always be of the same concrete type for all calls, and I run into object safe issues since rust sees that I'm trying to build a trait object from traits that are not object safe
Tldr til

**Review by wlmyng — APPROVED** [review]
🎉🎉🗣️🗣️
